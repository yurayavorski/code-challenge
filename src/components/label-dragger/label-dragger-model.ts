export enum ItemColors {
    Red = 'red',
    Blue = 'blue',
    Green = 'green',
    Yellow = ' yellow',
    White = 'white'
}

export interface ItemPosition {
    x: number
    y: number
}

export interface ItemState {
    isRendered: boolean
    position?: ItemPosition
    color: ItemColors
}
