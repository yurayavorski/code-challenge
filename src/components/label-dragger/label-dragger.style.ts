import styled from 'styled-components'
import { white } from '../../styles/colors'

export const DraggableItem = styled.div`
    position: absolute;
    display: inline-block;
    font-size: 42px;
    text-align: center;
    transition: color 0.3s;
    & label {
        border-radius: 5px;
        cursor: move;
        padding: 3px;
        transition: background 0.3s;
    }

    & > span {
        background: ${white};
        color: #000;
        font-size: 14px;
        padding: 5px;
        border-radius: 5px;
        display: block;
        position: absolute;
        text-align: center;
        width: 50px;
        top: 115%;
        left: 50%;
        margin-left: -30px;
        transition: 0.5s;
        & span {
            text-align: left;
            display: inline-block;
        }

        &:before {
            content: '';
            width: 0;
            height: 0;
            border-left: 8px solid transparent;
            border-right: 8px solid transparent;
            border-bottom: 8px solid ${white};
            position: absolute;
            top: -8px;
            left: 38%;
        }

        &.hidden {
            opacity: 0;
            visibility: hidden;
        }
    }
`
