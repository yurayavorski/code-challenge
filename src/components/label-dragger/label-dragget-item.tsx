import React, { useState } from 'react'
import { ItemState } from './label-dragger-model'
import Draggable, { DraggableData } from 'react-draggable'
import { DraggableItem } from './label-dragger.style'

export type OnDrag = (newPosition: DraggableData) => void

export interface LabelDraggerItemProps {
    itemState?: ItemState
    onDrag: OnDrag
    onStop: () => void
    onClick: () => void
    children: React.ReactNode
}

const HOVER_TIME = 2000

export const LabelDraggerItem = ({ itemState, onClick, onStop, onDrag, children }: LabelDraggerItemProps) => {
    const [showCoordinates, setShowCoordinates] = useState<boolean>(false)
    let timerId: number

    const onMouseOver = () => {
        timerId = setTimeout(() => setShowCoordinates(true), HOVER_TIME)
    }

    const interruptHover = () => {
        if (timerId) {
            clearTimeout(timerId)
        }
        setShowCoordinates(false)
    }

    return (
        <Draggable
            axis='both'
            handle='.handle'
            bounds='parent'
            defaultPosition={itemState?.position}
            position={itemState?.position}
            onDrag={(_, newPosition) => {
                interruptHover()
                onDrag(newPosition)
            }}
            onStop={onStop}
            grid={[1, 1]}
            scale={1}
        >
            <DraggableItem>
                <label
                    className='handle'
                    onClick={() => {
                        interruptHover()
                        onClick()
                    }}
                    onMouseOver={onMouseOver}
                    onMouseLeave={interruptHover}
                    style={{ color: itemState?.color }}
                >
                    {children}
                </label>
                <span className={!showCoordinates ? 'hidden' : ''}>
                    <span>
                        {`x: ${itemState?.position?.x}`}
                        <br />
                        {`y: ${itemState?.position?.y}`}
                    </span>
                </span>
            </DraggableItem>
        </Draggable>
    )
}
