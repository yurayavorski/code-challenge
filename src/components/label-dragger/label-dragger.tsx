import React, { useState } from 'react'
import { DraggableData } from 'react-draggable'
import { ItemColors, ItemState } from './label-dragger-model'
import { LabelDraggerItem, OnDrag } from './label-dragget-item'

export interface LabelDraggerProps {
    words: string[]
    itemStates?: ItemState[]
    setItemStates: React.Dispatch<React.SetStateAction<ItemState[] | undefined>>
}

export const getNextColor = (currentColor: ItemColors): ItemColors => {
    const colors = Object.values(ItemColors)
    const nextColorIndex = (colors.findIndex((x) => x === currentColor) + 1) % colors.length
    return colors[nextColorIndex]
}

export const LabelDragger = ({ words, itemStates, setItemStates }: LabelDraggerProps) => {
    const [isInDragState, setIsInDragState] = useState<boolean>(false)

    const onDrag = (index: number): OnDrag => (newPosition: DraggableData) => {
        setIsInDragState(true)
        const newItemStates = itemStates && [...itemStates]
        const itemPosition = newItemStates?.[index]?.position

        if (itemPosition) {
            itemPosition.x = newPosition.x
            itemPosition.y = newPosition.y

            setItemStates(newItemStates)
        }
    }

    const onStop = () => {
        setTimeout(() => setIsInDragState(false)) // Make sure onStop event is fired after the click event
    }

    const onItemClick = (index: number): void => {
        if (isInDragState) {
            return
        }

        const currentColor = itemStates?.[index].color
        if (itemStates && currentColor) {
            const nextColor = getNextColor(currentColor)
            const newState = [...itemStates]
            newState[index].color = nextColor

            setItemStates(newState)
        }
    }

    return (
        <>
            {words.length &&
                words.map(
                    (x, index) =>
                        itemStates &&
                        itemStates[index].isRendered && (
                            <LabelDraggerItem
                                key={index}
                                itemState={itemStates?.[index]}
                                onDrag={onDrag(index)}
                                onStop={onStop}
                                onClick={() => onItemClick(index)}
                            >
                                {x}
                            </LabelDraggerItem>
                        )
                )}
        </>
    )
}
