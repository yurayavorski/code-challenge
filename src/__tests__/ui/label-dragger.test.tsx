import React, { useState } from 'react'
import { cleanup, fireEvent, render } from '@testing-library/react'
import { getNextColor, LabelDragger } from '../../components/label-dragger/label-dragger'
import { ItemColors, ItemPosition, ItemState } from '../../components/label-dragger/label-dragger-model'
import drag from './test-utls/drag'
import { act } from 'react-dom/test-utils'
import { sleep } from './test-utls/sleep'

const testWordsSet = ['word1', 'word2']

const initialColor = ItemColors.Red

const LabelDraggerWrapper = () => {
    const [itemStates, setItemStates] = useState<ItemState[] | undefined>(
        testWordsSet.map(() => ({
            color: initialColor,
            isRendered: true,
            position: { x: 0, y: 0 }
        }))
    )

    return <LabelDragger words={testWordsSet} itemStates={itemStates} setItemStates={setItemStates} />
}

describe('Label Dragger', () => {
    afterEach(cleanup)

    it('renders label draggers with words', async () => {
        const { getByText, asFragment } = render(<LabelDraggerWrapper />)

        const initialRender = asFragment()
        expect(initialRender).toMatchSnapshot()
        expect(getByText(testWordsSet[0])).toBeTruthy()
        expect(getByText(testWordsSet[1])).toBeTruthy()
    })

    it('changes color of label after its clicked', async () => {
        const { getByText, asFragment, queryByText } = render(<LabelDraggerWrapper />)

        fireEvent.click(getByText(testWordsSet[0]))
        expect(queryByText(testWordsSet[0])?.style.color).toEqual(getNextColor(initialColor))
        const RenderAfterColorChanged = asFragment()
        expect(RenderAfterColorChanged).toMatchSnapshot()
    })

    it('changes position of element after its dragged', async () => {
        const { asFragment, queryByText } = render(<LabelDraggerWrapper />)

        const newItemPosition: ItemPosition = {
            x: 100,
            y: 100
        }

        await new Promise<void>((res) => {
            act(() => {
                drag(queryByText(testWordsSet[0]) as HTMLElement, {
                    to: newItemPosition,
                    delta: newItemPosition,
                    duration: 500,
                    steps: 20
                }).then(() => res())
            })
        })

        const RenderAfterDrag = asFragment()
        expect(RenderAfterDrag).toMatchSnapshot()
    })

    it('shows the tooltip after 2000ms hover', async () => {
        const { asFragment, queryByText } = render(<LabelDraggerWrapper />)

        act(() => {
            fireEvent.mouseOver(queryByText(testWordsSet[0]) as HTMLElement)
        })

        await sleep(2100)

        const RenderAfterHover = asFragment()
        expect(RenderAfterHover).toMatchSnapshot()
    })
})
