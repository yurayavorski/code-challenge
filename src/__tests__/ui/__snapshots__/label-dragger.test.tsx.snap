// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label Dragger changes color of label after its clicked 1`] = `
<DocumentFragment>
  <div
    class="sc-bdVaJa eyzjMJ react-draggable"
    style="transform: translate(0px,0px);"
  >
    <label
      class="handle"
      style="color: blue;"
    >
      word1
    </label>
    <span
      class="hidden"
    >
      <span>
        x: 0
        <br />
        y: 0
      </span>
    </span>
  </div>
  <div
    class="sc-bdVaJa eyzjMJ react-draggable"
    style="transform: translate(0px,0px);"
  >
    <label
      class="handle"
      style="color: red;"
    >
      word2
    </label>
    <span
      class="hidden"
    >
      <span>
        x: 0
        <br />
        y: 0
      </span>
    </span>
  </div>
</DocumentFragment>
`;

exports[`Label Dragger changes position of element after its dragged 1`] = `
<DocumentFragment>
  <div
    class="sc-bdVaJa eyzjMJ react-draggable react-draggable-dragged"
    style="transform: translate(100px,100px);"
  >
    <label
      class="handle"
      style="color: red;"
    >
      word1
    </label>
    <span
      class="hidden"
    >
      <span>
        x: 100
        <br />
        y: 100
      </span>
    </span>
  </div>
  <div
    class="sc-bdVaJa eyzjMJ react-draggable"
    style="transform: translate(0px,0px);"
  >
    <label
      class="handle"
      style="color: red;"
    >
      word2
    </label>
    <span
      class="hidden"
    >
      <span>
        x: 0
        <br />
        y: 0
      </span>
    </span>
  </div>
</DocumentFragment>
`;

exports[`Label Dragger renders label draggers with words 1`] = `
<DocumentFragment>
  <div
    class="sc-bdVaJa eyzjMJ react-draggable"
    style="transform: translate(0px,0px);"
  >
    <label
      class="handle"
      style="color: red;"
    >
      word1
    </label>
    <span
      class="hidden"
    >
      <span>
        x: 0
        <br />
        y: 0
      </span>
    </span>
  </div>
  <div
    class="sc-bdVaJa eyzjMJ react-draggable"
    style="transform: translate(0px,0px);"
  >
    <label
      class="handle"
      style="color: red;"
    >
      word2
    </label>
    <span
      class="hidden"
    >
      <span>
        x: 0
        <br />
        y: 0
      </span>
    </span>
  </div>
</DocumentFragment>
`;

exports[`Label Dragger shows the tooltip after 2000ms hover 1`] = `
<DocumentFragment>
  <div
    class="sc-bdVaJa eyzjMJ react-draggable"
    style="transform: translate(0px,0px);"
  >
    <label
      class="handle"
      style="color: red;"
    >
      word1
    </label>
    <span
      class=""
    >
      <span>
        x: 0
        <br />
        y: 0
      </span>
    </span>
  </div>
  <div
    class="sc-bdVaJa eyzjMJ react-draggable"
    style="transform: translate(0px,0px);"
  >
    <label
      class="handle"
      style="color: red;"
    >
      word2
    </label>
    <span
      class="hidden"
    >
      <span>
        x: 0
        <br />
        y: 0
      </span>
    </span>
  </div>
</DocumentFragment>
`;
